src/register.php

```diff
@@ -33,34 +34,23 @@
 			// TODO: remove this column from the database
 			$IP = 0;
 
-			$sql = "SELECT email FROM pending_validations WHERE email=?";
-			$rows = [];
-			if ( $sth = $dice->dbconn->prepare( $sql )) {
-				$sth->bind_param('s',$email);
-				$sth->execute() or trigger_error($mysqli->error);
-				$sth->bind_result($emailColumn);
-				while($sth->fetch()) {
-					$rows[] = $emailColumn;
-				}
-			} else {
-				echo "A DB error has occured, please contact an admin. (";
-				var_dump( $dice->dbconn->errno );
-				echo ")";
-				exit;
-			}
+			$sth = $dice->dbconn->prepare("SELECT email FROM pending_validations WHERE email=?");
+			$sth->bind_param('s', $email);
+			$sth->execute() or trigger_error($mysqli->error);
 
 			// insert or update pending validation
-			if (empty($rows)) {
+			if ($sth->num_rows) {
 				$sql = "INSERT INTO pending_validations (email, validation_key, time_stamp, IP) VALUES (?, ?, FROM_UNIXTIME(?), ?)";
-				$sth = $dice->dbconn->prepare( $sql );
-				$sth->bind_param('ssss', $email, $validation, $time, $IP );
+				$sth = $dice->dbconn->prepare($sql);
+				$sth->bind_param('ssss', $email, $validation, $time, $IP);
 				$sth->execute();
 			} else {
-				$sql = "UPDATE pending_validations SET validation_key=?, time_stamp= FROM_UNIXTIME(?), IP=? WHERE email=?";
-				$sth = $dice->dbconn->prepare( $sql );
-				$sth->bind_param('ssss', $validation, $time, $IP, $email );
+				$sql = "UPDATE pending_validations SET validation_key=?, time_stamp=FROM_UNIXTIME(?), IP=? WHERE email=?";
+				$sth = $dice->dbconn->prepare($sql);
+				$sth->bind_param('ssss', $validation, $time, $IP, $email);
```

Comment:
When I run the code as-is, I'm seeing the following error:

```
NOTICE: PHP message: PHP Fatal error:  Uncaught Error: Call to a member function bind_param() on boolean in /usr/share/nginx/html/dice/register.php:50
```

I think this corroborates the comment above that the expression using `$sth->num_rows` needs to be compared to 0.  However, even with that, I'm not sure why the `prepare()` call would fail; I would have expected the error to occur when the SQL was executed (or just a logic error with 0 rows updated).  `prepare()` failing suggests that the SQL is malformed, but it's basically the same as before, unless my eyes are too tired to see something.

Reply:
One reason for the `prepare()` error may be the removal of the `mysqli_select_db()` call, which selected the default database.  (I can see `prepare()` failing if the table/column names are not valid in the default database.)  I would have expected the default database to be selected by the fourth parameter to the `mysqli()` constructor, but maybe that's not the case?

Reply:
> One reason for the prepare() error may be the removal of the mysqli_select_db() call, which selected the default database. (I can see prepare() failing if the table/column names are not valid in the default database.) I would have expected the default database to be selected by the fourth parameter to the mysqli() constructor, but maybe that's not the case?

Pretty sure this isn't the problem. Following this logic the previous checkIfEmail is registered call would fail as well.

In case I can't get docker to work tomorrow, could you try getting a more descriptive error message using `$dice->dbconn->error_list` and `$dice->dbconn->errno`?

Reply:
Here's the change I made:

```diff
diff --git a/src/register.php b/src/register.php
index 4839cb5..7f49b38 100644
--- a/src/register.php
+++ b/src/register.php
@@ -39,9 +39,15 @@
                        $sth->execute() or trigger_error($mysqli->error);
 
                        // insert or update pending validation
-                       if ($sth->num_rows) {
+                       if ($sth->num_rows === 0) {
                                $sql = "INSERT INTO pending_validations (email, validation_key, time_stamp, IP) VALUES (?, ?, FROM_UNIXTIME(?), ?)";
                                $sth = $dice->dbconn->prepare($sql);
+                               error_log("sth type = " . gettype($sth));
+                               if (!$sth) {
+                                       error_log("errno = " . $dice->dbconn->errno);
+                                       error_log("error = " . $dice->dbconn->error);
+                                       error_log("error_list = " . implode(",", $dice->dbconn->error_list));
+                               }
                                $sth->bind_param('ssss', $email, $validation, $time, $IP);
                                $sth->execute();
                        } else {
```

And here are the resulting logs during a registration:

```
app_1  | [07-Jan-2018 04:00:36] WARNING: [pool www] child 7 said into stderr: "NOTICE: PHP message: sth type = boolean"
app_1  | [07-Jan-2018 04:00:36] WARNING: [pool www] child 7 said into stderr: "NOTICE: PHP message: errno = 0"
app_1  | [07-Jan-2018 04:00:36] WARNING: [pool www] child 7 said into stderr: "NOTICE: PHP message: error = "
app_1  | [07-Jan-2018 04:00:36] WARNING: [pool www] child 7 said into stderr: "NOTICE: PHP message: error_list = "
app_1  | [07-Jan-2018 04:00:36] WARNING: [pool www] child 7 said into stderr: "NOTICE: PHP message: PHP Fatal error:  Uncaught Error: Call to a member function bind_param() on boolean in /usr/share/nginx/html/dice/register.php:51"
app_1  | [07-Jan-2018 04:00:36] WARNING: [pool www] child 7 said into stderr: "Stack trace:"
app_1  | [07-Jan-2018 04:00:36] WARNING: [pool www] child 7 said into stderr: "#0 {main}"
app_1  | [07-Jan-2018 04:00:36] WARNING: [pool www] child 7 said into stderr: "  thrown in /usr/share/nginx/html/dice/register.php on line 51"
app_1  | 172.18.0.4 -  07/Jan/2018:04:00:36 +0000 "POST /register.php" 200
```

`prepare()` returns `FALSE` but no errors are reported.  Awesome!

Reply:
Ok, there may be some kind of variable scope issue going on here that I don't understand about PHP.  I assigned the result of the second `prepare()` call to a new variable, `$sth2`.  When I do that, I now get an actual error message in the logs:

```
app_1  | [07-Jan-2018 04:21:28] WARNING: [pool www] child 9 said into stderr: "NOTICE: PHP message: sth2 type = boolean"
app_1  | [07-Jan-2018 04:21:28] WARNING: [pool www] child 9 said into stderr: "NOTICE: PHP message: errno = 2014"
app_1  | [07-Jan-2018 04:21:28] WARNING: [pool www] child 9 said into stderr: "NOTICE: PHP message: error = Commands out of sync; you can't run this command now"
app_1  | [07-Jan-2018 04:21:28] WARNING: [pool www] child 9 said into stderr: "NOTICE: PHP message: error_list = "
```

The root cause of the out-of-sync error seems to be due to the first statement not being closed before creating the second statement.  This change appears to allow registration to succeed:

```diff
diff --git a/src/register.php b/src/register.php
index 4839cb5..a9ec3f8 100644
--- a/src/register.php
+++ b/src/register.php
@@ -37,9 +37,11 @@
                        $sth = $dice->dbconn->prepare("SELECT email FROM pending_validations WHERE email=?");
                        $sth->bind_param('s', $email);
                        $sth->execute() or trigger_error($mysqli->error);
+                       $isNewRegistration = $sth->num_rows === 0;
+                       $sth->close();
 
                        // insert or update pending validation
-                       if ($sth->num_rows) {
+                       if ($isNewRegistration) {
                                $sql = "INSERT INTO pending_validations (email, validation_key, time_stamp, IP) VALUES (?, ?, FROM_UNIXTIME(?), ?)";
                                $sth = $dice->dbconn->prepare($sql);
                                $sth->bind_param('ssss', $email, $validation, $time, $IP);
```

However, when I attempt to validate the registration, validation fails.  It looks like the validation key in the DB does not match what was in the email link (I verified the record is there, it's just the `validation_key` column value doesn't match what was in the email.)

Reply:
@ssoloff I did some research on this issue and here's my theory:
The out of sync error happens whenever a query/prepared statement is getting created before the result of the previous query/prepared statement has been freed.
So because querying the number of affected rows doesn't "consume" the result in any way, mysqli still waits for the result to be processed. Closing the statement therefore correctly fixes this issue. The result is freed and the next query can be created.
So why didn't the results show up?
My theory is that because the first query still was active and was never closed mysqli->errno/error/error_list would display the errors of the first query, which had no errors.
I'm not sure why exactly changing the variable name to something different fixed the issue, but I'd assume it has something to do with the reference to the statement object that no longer exists, because it has been overridden with false and deconstructing the object at some point.

Reply:
Weird. all I can think of is [this change](https://github.com/triplea-game/dice-server/pull/30/files?w=1#diff-68022f3e8f99fe5b8bff52c00f520a1cL12), that could potentially strip off special characters and html tags, but other than that nothing should've changed really.
So the question really should be in what respect the token really differs.

Reply:
> My theory is that because the first query still was active and was never closed

I'm not familiar with how garbage collection works in PHP, but this theory makes sense if it is eager and not deferred as in Java.

> So the question really should be in what respect the token really differs.

I'll re-run another test this evening and post the results.  I couldn't figure it out last night via just a code review because it appears the `$validation` variable in _register.php_ was used, without modification, in both the `INSERT` statement and the email link.  I don't think it would have anything to do with _validate.php_ because I literally saw a different value in the email vs. the database (not just what value the _validate.php_ code was operating on).

Reply:
@ssoloff Thanks, I really appreciate this.
The only explanation I could think of would be that a value of an old registration was left in the db, and the code didn't properly overwrite this or that the validation check isn't working and not properly deleting the value from the database either.

Reply:
The test was performed on a fresh Docker container, so the DB should have been empty at the time, but I'll verify.

Reply:
@RoiEXLab I re-ran the test with the latest code from this branch.  I may have been mistaken earlier when I said the validation keys were different (maybe I was looking at the wrong email?).  In my most recent test, they are identical, so that's good.  However, I still receive the message, "Could not verify the data. Please check the link you have received in your email."  I'll try to dig deeper in a bit.

Reply:
> The only explanation I could think of would be that a value of an old registration was left in the db

@RoiEXLab So, after more testing, I now believe this is correct.  I must not have recreated the MySQL container, as I thought I did. 😞  It looks like `$isNewRegistration` is always set to `TRUE` even if the email is already present in the `pending_validations` table, which skips the `UPDATE` branch.

The root cause for this problem, as well as the email validation problem, appears to be due to not reading the entire recordset before checking `$sth->num_rows`.  I had to call `$sth->store_result()` to force the recordset to be read to get the correct number of rows returned in each case:

```diff
diff --git a/src/register.php b/src/register.php
index 5dadee7..cf8de9e 100644
--- a/src/register.php
+++ b/src/register.php
@@ -37,6 +37,7 @@
                        $sth = $dice->dbconn->prepare("SELECT email FROM pending_validations WHERE email=?");
                        $sth->bind_param('s', $email);
                        $sth->execute() or trigger_error($mysqli->error);
+                       $sth->store_result();
 
                        $isNewRegistration = $sth->num_rows === 0;
                        $sth->close();
diff --git a/src/validate.php b/src/validate.php
index 66c74a8..dee70c3 100644
--- a/src/validate.php
+++ b/src/validate.php
@@ -18,6 +18,7 @@
                $sth = $dice->dbconn->prepare("SELECT email FROM pending_validations WHERE email=? AND validation_key=?");
                $sth->bind_param('ss', $email, $validation);
                $sth->execute() or trigger_error($mysqli->error);
+               $sth->store_result();
                if ($sth->num_rows === 0) {
                        exit("Could not verify the data. Please check the link you have received in your email");
                }
```

Would it be better to change these queries to a `SELECT COUNT(*) ...` and just use `bind_param_result()` to store the result in a local variable instead of transferring all the rows and counting them?

**NOTE:** I would suggest checking other usages of `num_rows` to ensure we're correctly reading the entire recordset in those locations.

Reply:
Thanks for digging into this, I have an even better idea for the register.php file.
But will probably fall back to your Count suggestion on all the other usages 👍